include/circt/Dialect/HW/HWAttributesNaming.td

```diff
@@ -17,32 +17,71 @@ include "circt/Dialect/HW/HWDialect.td"
+    This works like a symbol reference, but to a name inside a list of `InnerSymbolTable`'s.
   }];
   let mnemonic = "innerNameRef";
-  let parameters = (ins "::mlir::FlatSymbolRefAttr":$moduleRef,
-                        "::mlir::StringAttr":$name);
+  let parameters = (ins ArrayRefParameter<"::mlir::StringAttr">:$path);
 
   let builders = [
-    AttrBuilderWithInferredContext<(ins "::mlir::StringAttr":$module,
-                                        "::mlir::StringAttr":$name), [{
-      return $_get(
-          module.getContext(), mlir::FlatSymbolRefAttr::get(module), name);
-    }]>
+
+    AttrBuilderWithInferredContext<(ins "::mlir::ArrayRef<::mlir::StringAttr>":$pathAttr), [{
+      assert(!pathAttr.empty() && "Empty path");
+      assert(pathAttr.front() && "Null path element");
+      return InnerRefAttr::get(pathAttr.front().getContext(), pathAttr);
+    }]>,
+  
+    AttrBuilderWithInferredContext<(ins "::mlir::ArrayRef<::mlir::StringAttr>":$pathAttr,
+                                        "::mlir::StringAttr":$leaf), [{
+      llvm::SmallVector<::mlir::StringAttr> fullPath(pathAttr.begin(), pathAttr.end());
+      fullPath.push_back(leaf);
+      return InnerRefAttr::get(leaf.getContext(), fullPath);
+    }]>,
+
+    // @mortbopet: Old InnerRef single-module compatability builders.
+    AttrBuilderWithInferredContext<(ins "::mlir::StringAttr":$leaf), [{
+      llvm::SmallVector<::mlir::StringAttr> fullPath;
+      fullPath.push_back(leaf);
+      return InnerRefAttr::get(fullPath);
+    }]>,
+
+    AttrBuilderWithInferredContext<(ins 
+                                    "::mlir::StringAttr":$parent,
+                                    "::mlir::StringAttr":$leaf), [{
+      ::llvm::SmallVector<::mlir::StringAttr> fullPath{parent};
+      fullPath.push_back(leaf);
+      return InnerRefAttr::get(fullPath);
+    }]>,
+    AttrBuilder<(ins "::mlir::StringAttr":$parent,
+                     "::mlir::StringAttr":$leaf), [{
+      return InnerRefAttr::get(parent, leaf);
+    }]>,
   ];
 
   let hasCustomAssemblyFormat = 1;
+  let genVerifyDecl = 1;
 
   let extraClassDeclaration = [{
-    /// Get the InnerRefAttr for an operation and add the sym on it.
-    static InnerRefAttr getFromOperation(mlir::Operation *op,
-                                         mlir::StringAttr symName,
-                                         mlir::StringAttr moduleName);
+    ::mlir::StringAttr getRootReference() const {
+      assert(!getPath().empty() && "Empty path");
+      return getPath().front();
+    }
+    ::mlir::StringAttr getLeafReference() const {
+      assert(!getPath().empty() && "Empty path");
+      return getPath().back();
+    }
+    ::mlir::StringAttr getName() const {
+      return getLeafReference();
+    }
 
-    /// Return the name of the referenced module.
-    mlir::StringAttr getModule() const { return getModuleRef().getAttr(); }
+    // Old InnerRefAttr compatability accessors
```

Comment:
Why have these?

Reply:
Mostly to reduce the footprint of this PR. Can remove them as part of this PR if reviewers are willing to accept the increase in PR scope.